ENIGMAsystem/SHELL/Universal_System/Extensions/FileDropper/FileDropper.cpp

```diff
@@ -0,0 +1,184 @@
+static HDROP hDrop = NULL;
+static string fname;
+
+static string def_pattern;
+static bool def_allowfiles = true;
+static bool def_allowdirs = true;
+static bool def_allowmulti = true;
+
+using enigma_user::filename_name;
+using enigma_user::filename_ext;
+using enigma_user::file_exists;
+using enigma_user::directory_exists;
+
+static void UnInstallHook(HHOOK Hook) {
+  UnhookWindowsHookEx(Hook);
+}
+
+static LRESULT CALLBACK HookWndProc(HWND hWnd, UINT uMsg, WPARAM wParam, LPARAM lParam) {
+  LRESULT rc = CallWindowProc(oldProc, hWnd, uMsg, wParam, lParam);
+
+  if (uMsg == WM_DROPFILES) {
+    if (!def_allowmulti) fname = "";
+    hDrop = (HDROP)wParam;
+
+    UINT nNumOfFiles = DragQueryFileW(hDrop, 0xFFFFFFFF, NULL, 0);
+    if (nNumOfFiles > 0) {
+      for (UINT i = 0; i < nNumOfFiles; i += 1) {
+        UINT nBufSize = DragQueryFileW(hDrop, i, NULL, 0) + 1;
+        wchar_t *fName = new wchar_t[nBufSize];
+        DragQueryFileW(hDrop, i, fName, nBufSize);
+        if (fname != "") fname += "\n";
+        fname += shorten(fName);
+        delete[] fName;
+      }
+    }
+
+    std::vector<string> nameVec = split_string(fname, '\n');
+    sort(nameVec.begin(), nameVec.end());
+    nameVec.erase(unique(nameVec.begin(), nameVec.end()), nameVec.end());
+    fname = "";
+
+    std::vector<string>::size_type sz = nameVec.size();
+    for (std::vector<string>::size_type i = 0; i < sz; i += 1) {
+      if (fname != "") fname += "\n";
+      fname += nameVec[i];
+    }
+
+    DragFinish(hDrop);
+  }
+
+  return rc;
+}
+
+static LRESULT CALLBACK SetHook(int nCode, WPARAM wParam, LPARAM lParam) {
+  if (nCode == HC_ACTION) {
+    CWPSTRUCT *pwp = (CWPSTRUCT *)lParam;
+
+    if (pwp->message == WM_KILLFOCUS) {
+      oldProc = (WNDPROC)SetWindowLongPtrW(enigma::hWnd, GWLP_WNDPROC, (LONG_PTR)HookWndProc);
+      UnInstallHook(hook);
+    }
+  }
+
+  return CallNextHookEx(hook, nCode, wParam, lParam);
+}
+
+static HHOOK InstallHook() {
+  hook = SetWindowsHookExW(WH_CALLWNDPROC, (HOOKPROC)SetHook, NULL, GetWindowThreadProcessId(enigma::hWnd, NULL));
+  return hook;
+}
+
+static void file_dnd_apply_filter(string pattern, bool allowfiles, bool allowdirs, bool allowmulti) {
+  if (pattern == "") { pattern = "."; }
+  pattern = string_replace_all(pattern, " ", "");
+  pattern = string_replace_all(pattern, "*", "");
+  std::vector<string> extVec = split_string(pattern, ';');
+  std::vector<string> nameVec = split_string(fname, '\n');
+  std::vector<string>::size_type sz1 = nameVec.size();
+  std::vector<string>::size_type sz2 = extVec.size();
+  fname = "";
+
+  for (std::vector<string>::size_type i2 = 0; i2 < sz2; i2 += 1) {
+    for (std::vector<string>::size_type i1 = 0; i1 < sz1; i1 += 1) {
+      if (extVec[i2] == "." || extVec[i2] == filename_ext(nameVec[i1])) {
+        if (fname != "") fname += "\n";
+        fname += nameVec[i1];
+      }
+    }
+  }
+
+  nameVec = split_string(fname, '\n');
+  sz1 = nameVec.size();
+  fname = "";
```

Comment:
It's inefficient to repeatedly serialize and deserealize the data you're working with.